### PR TITLE
Add proposed output def change

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/annotated/sample.py [hunk 1]
```diff
@@ -1,13 +1,14 @@
 from flytekit.annotated.stuff import task, workflow, WorkflowOutputs
+from flytekit.annotated.type_engine import outputs
 
 
-@task(outputs=['s_out'])
-def x(s: int) -> int:
+@task
+def x(s: int) -> outputs(s_out=int):
```
Comment (kumare3) [inline flytekit/annotated/sample.py:6]: tbh I do not like this, because this needs the user to change the interface of their tasks :(

Comment (EngHabu) [reply flytekit/annotated/sample.py:6]: Can you elaborate? Change the interface of current flytekit style tasks? Or of py3 style functions?
This is, from python perspective, equivalent to `(int, str)`

Comment (EngHabu) [reply flytekit/annotated/sample.py:6]: And if you return a single output, you don't need to do this and we will automatically name it just "output"

File: tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    
+#     assert my_task.unit_test() == {'output': 'Hello world'}
+
+
+@flyte_test
+def test_read_file_known_location():
+    # Option 1
+    # users receive a file handle as the parameter to their task
+    @python_task
+    def my_task(ctx, fh: typing.BinaryIO):
```
Comment (kumare3) [inline tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py:53]: Love this!

File: tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        with open(file_path.local_path, mode='rb') as fh:
+            lines = fh.readlines()
+            # assert 
+
+    assert my_task.unit_test(file_path=CustomPathLike('s3://bucket-my/mytest')) == {}
+
+    # Option 2.2
+    # Provide a custom open() that encapsulates download and open actions
+    @python_task
+    def my_task(ctx, file_path: CustomPathLike):
```
Comment (kumare3) [inline tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py:83]: this is standard, so love this

File: tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        with file_path.open() as fh:
+            lines = fh.readlines()
+            # assert 
+
+
+    # Option 2.3
+    # Flyte always downloads the file and users can use the regular open() function to read it.
+    @python_task
+    def my_task(ctx, file_path: CustomPathLike):
```
Comment (kumare3) [inline tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py:92]: i actually dont want to assume this, as this has perf implications

File: tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    # Option 1.2
+    # Users pass a Path-Like object that flyte knows how to interpret and open or youo
+    assert my_task.unit_test(fh=CustomPathLike('/mytest', format="csv")) == {}
+    
+    # Option 2
+    # Users receive a Path-Like type as a parameter to their function (much like how Blobs work today)
+    # Option 2.1
+    # Users will have to call downlooad then open
+    @python_task
+    def my_task(ctx, file_path: CustomPathLike):
```
Comment (kumare3) [inline tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py:72]: i think this is an overhead

File: tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+from io import BufferedWriter
+
+from flytekit.sdk.test_utils import flyte_test
+from flytekit.sdk.tasks import python_task
+from flytekit.annotated.type_engine import outputs
+
+@flyte_test
+def test_simple_input_output():
+    @python_task
+    def my_task(ctx, a: int) -> outputs(b=int, c=str):
```
Comment (kumare3) [inline tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py:12]: do we need to wrap with `outputs()`
I think this is more intrusive?

File: flytekit/annotated/sample.py (hunk 1 above)
Comment (kumare3) [reply flytekit/annotated/sample.py:6]: What I mean, is that they have to have an explicit dependency on 
```python
from flytekit. ... import outputs
#then use outputs
```

What I want is ideally just annotate with one thing and you are done, if you remove the annotation, your code works as is?
Maybe unreasonable

File: tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+#     assert my_task.unit_test(fh=CustomPathLike('/mytest', format="csv")) == {}
+    
+#     # Option 2
+#     # Users receive a Path-Like type as a parameter to their function (much like how Blobs work today)
+#     # Option 2.1
+#     # Users will have to call downlooad then open
+#     @python_task
+#     def my_task(ctx, file_path: CustomPathLike):
+#         file_path.download()
+#         with open(file_path.local_path, mode='rb') as fh:
```
Comment (wild-endeavor) [inline tests/flytekit/unit/use_scenarios/unit_testing/test_type_hints.py:89]: if this is how you open the locally downloaded file, what would `print(file_path)` show?  I feel like it should be the remote path.  That is, either there's a `remote_path` or there's a `local_path` but not both, and the naked object should show the other one.

File: flytekit/sdk/tasks.py [hunk 8]
```diff
@@ -16,6 +16,16 @@
 from flytekit.sdk.spark_types import SparkType as _spark_type
 
 
+def current_context() -> _sdk_runnable_tasks.ExecutionParameters:
```
Comment (wild-endeavor) [inline flytekit/sdk/tasks.py:19]: I want to move this into the engine.

File: flytekit/common/utils.py [hunk 9]
```diff
@@ -205,6 +207,30 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         return False
 
 
+T = typing.TypeVar('T')
+
+
+class StackableContext(typing.Generic[T]):
```
Comment (wild-endeavor) [inline flytekit/common/utils.py:213]: why is this a generic?

**Pull request merged**